src/dash-table/utils/applyClipboardToData.ts

```diff
@@ -23,29 +24,41 @@ export default (
     }
 
     // don't modify the data and columns directly -- we may abort the paste
+    // Individual rows will be modified, needs to be a deep clone
     let newData = R.clone(data);
-    const newColumns = R.clone(columns);
+    // Might add columns, not modifying the columns themselves, shallow clone is sufficient
+    let newColumns = columns_.slice(0);
+    let newVisibleColumns = visibleColumns.slice(0);
 
-    if (overflowColumns && values[0].length + (activeCell as any).column >= columns.length) {
+    if (overflowColumns && values[0].length + (activeCell as any).column >= visibleColumns.length) {
+        const _newColumns = [];
         for (
-            let i = columns.length;
+            let i = visibleColumns.length;
             i < values[0].length + (activeCell as any).column;
             i++
         ) {
-            newColumns.push({
+            _newColumns.push({
                 id: `Column ${i + 1}`,
                 name: `Column ${i + 1}`,
                 type: ColumnType.Any,
                 sort_as_null: []
             } as any);
             newData.forEach(row => (row[`Column ${i}`] = ''));
         }
+
+        newColumns = R.insertAll(
+            R.indexOf(R.last(visibleColumns), columns_) + 1,
+            _newColumns,
+            newColumns
+        );
+
+        newVisibleColumns = R.concat(newVisibleColumns, _newColumns);
```

Comment:
When new columns are created, they are appended after the last visible column. Both visible and "all" columns are updated to both (1) handle the copy/paste below and (2) the columns update correctly.

Reply:
hmm OK - I might have added the new columns after trailing hidden columns as well, but it's probably impossible to declare one way objectively better than the other. This is fine.